xls/dslx/run_routines.cc

```diff
@@ -258,6 +258,8 @@ absl::StatusOr<TestResult> ParseAndTest(std::string_view program,
 
   auto handle_error = [&](const absl::Status& status,
                           std::string_view test_name, bool is_quickcheck) {
+    if (!options.handle_status)
```

Comment:
can you add more context in the PR description about why this change is necessary?

just looking at the change, it's not clear to me why you wouldn't want to always return the status to the caller.

Reply:
With this option I was able to keep default behavior of ParseAndTest.

ParseAndTest would return on first failed test, if it would return test status. 
Right now, it logs cause of the failure and continues to execute remaining tests.

Reply:
ah, so this flag is mostly there to verify that the test case that reproduce the bug fails early?

if yes, maybe we should modify the the assertion so that they ignore other test results?

Reply:
I don't think it's a good idea to modify assertion behavior.
What I see right now, is that, all DSLX tests use assertions to check functions/proc outputs.
Early escape on assertion failure would bring us back to skipping tests after first fail.

Reply:
> Early escape on assertion failure would bring us back to skipping tests after first fail.

But making this behavior configurable (being able to basically `fail_fast`) seems orthogonal to what this PR is fixing (the type info propagation), do you think it's worth splitting in a separate issue/pr?

Reply:
> I can't create tests for assert_* without fast_fail

couldn't you extract the number of failure and the condition without `fast_fail`?

Reply:
I'm unable to do so from ParseAndTest output in its unmodified version. 
https://github.com/google/xls/blob/d28e813f1d56c8ec1c3502219e14b48b8a82be4f/xls/dslx/run_routines.cc#L371-L375
https://github.com/google/xls/blob/d28e813f1d56c8ec1c3502219e14b48b8a82be4f/xls/dslx/run_routines.cc#L259-L282
Captures all interpreter errors.

Reply:
Yes, but since you're the one creating the `test_proc` you know they are only supposed to yield one error and could potentially assert this in the output?

Reply:
oh, so you mean that you're unable to get the test to fails with the previous version of the code?

Since your change does fix the type inference, I think what you have with checking that the error message includes "not less than" is enough.

In the passing case (with this patch applied) the error message with be surfaced, without the error would be lost and the test will fail.

Am I missing something?

Reply:
No, that's exactly what was and is happening there.
I hope now it’s clearer why I had to add this new flag.

Reply:
> error message is lost in ParseAndTest error handler

My understanding was that the error message was lost in ParseAndTest error handler because or the concrete type missing error this PR is fixing, or is that another issue?

Reply:
By saying "error message is lost" I meant that it isn't propagated to ParseAndTest caller.

Reply:
Yep, but it sounds that now that you're fixing the type deduction error, the correct message should be surfaced.

Are you worried that you wouldn't be able to identify a regression? (and that's why you're adding the flag?), would there be a downside to always returning the error from `ParseAndTest`?

Reply:
Before changes ParseAndTest would only return status that either indicated all tests passed or at least one failed, without additional error message. I can't distinguish between test failing on type deduction from test failing on assertion. I'm using status error message to distinguish between these 2 options and without it I can't detect regression.

I think we can always return `status` , I'll change it to do so.